financial_calculations/forward_curves.py

```diff
@@ -0,0 +1,74 @@
+    
+    Parameters:
+    -----------
+    cmt_data : list of tuples
+        Each tuple contains (maturity_years, coupon_rate) for bonds.
+        Coupon rate should be expressed as a decimal (e.g., 5% as 0.05).
+    market_close_date : datetime or datetime64[D]
+        The market close date from which the cash flows are discounted.
+    par_value : float
+        The par value of the bond.
+    initial_guess : float, optional
+        The initial guess for each spot rate, default is 0.03.
+        
+    Returns:
+    --------
+    tuple:
+        A tuple containing two elements:
+        - spot_rate_dates (np.ndarray): Array of spot rate dates.
+        - spot_rates (np.ndarray): Array of spot rates corresponding to each spot rate date.
+    
+    Raises:
+    -------
+    ValueError:
+        If minimization fails to converge for a specific bond.
+    """
+
+    # If the market close date is input as a datetime64[D] type, convert to datetime for relativedelta operations in the future
+    if isinstance(market_close_date, np.datetime64): 
+        market_close_date_dt = market_close_date.astype(datetime)
+        market_close_date = datetime.combine(market_close_date_dt, datetime.min.time()) # This adds the HMS 00:00:00 to the datetime object 
+            # which makes it identical to a datetime object initialized with the month, day, and year of the given numpy datetime64[D]
+        
+    spot_rate_dates = np.array([market_close_date])
+    spot_rates = np.array([])
+
+    for maturity_years, coupon in cmt_data:
+        # Append the spot rate date associated with the current maturity year
+        spot_rate_dates = np.append(spot_rate_dates, market_close_date + relativedelta(years=maturity_years))
+
+        # Generate cash flows for the bond
+        payment_dates, cash_flows = create_semi_bond_cash_flows(market_close_date, par_value, coupon, maturity_years)
+
+        # Objective function to minimize the squared difference between the bond price and the par value
+        def objective(rate: float):
+            discount_rates = np.append(spot_rates, rate) # Append the spot rate date associated with the current maturity year
+            price = discount_cash_flows(payment_dates, cash_flows, discount_rates, spot_rate_dates)  # Discount the cash flows using the new spot rate
+
+            return (price - par_value)**2 # Return the difference squared as the quantity to be minimized
+
+        # Minimize the objective function using a 'L-BFGS-B' method to find the best spot rate
+        result = minimize(objective, x0=initial_guess, method='L-BFGS-B', bounds=[(0, 1)], options={'ftol': 1e-4}) # We set a tolerance level to make sure the minimizer converges
+
+        if result.success:
+            spot_rates = np.append(spot_rates, result.x[0])
+        else:
+            raise ValueError(f"Minimization did not converge for payment date {payment_dates[-1]}.")
+
+    return spot_rate_dates, spot_rates
```

Comment:
is this actually a tuple?  legit question.  I actually have no idea.

Reply:
yeah I've returned as a tuple like this for cs ckasses